File: C++/567. Permutation in String.cpp
class Solution {

public:
    bool checkInclusion(string s1, string s2)
    {
        if (s2.size() < s1.size())
            return false;
        vector<int>hash(26, 0), check(26, 0);

        for (char i : s1)
            ++check[i - 'a'];

        int i = 0, j = 0;

        while (i < s1.size())
            ++hash[s2[i++] - 'a'];
        if (hash == check)
            return true;

        while (i < s2.size())
        {
            --hash[s2[j++] - 'a'];
            ++hash[s2[i++] - 'a'];
            if (hash == check)
                return true;
        }
        return false;
    }
};

File: Python/567. Permutation in String.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if len(s2) < len(s1):
            return False

        hash = [0] * 26
        check = [0] * 26

        for i in s1:
            check[ord(i) - ord('a')] += 1

        i = 0
        j = 0

        while i < len(s1):
            hash[ord(s2[i]) - ord('a')] += 1
            i += 1

        if hash == check:
            return True

        while i < len(s2):
            hash[ord(s2[j]) - ord('a')] -= 1
            j += 1
            hash[ord(s2[i]) - ord('a')] += 1
            i += 1

            if hash == check:
                return True

        return False